File: gateway.lua
-- gateway.lua
local MANIFEST_VERSION = 2

-- The human-readable config
local config = {
    topology = "_/_\\", -- Root with 2 leaves
    values = {
        { type = "string", val = "sensor_01" },
        { type = "float",  val = 25.3 } -- The original temperature
    }
}

-- Encode values to the ts_values_decode-compatible blob:
-- [u32 little-endian count] then count x [u32 little-endian length][bytes]
local blob = string.pack("<I4", 2) -- two values
local str = config.values[1].val
blob = blob .. string.pack("<I4", #str) .. str          -- value 0: length-prefixed string
blob = blob .. string.pack("<I4", 4) .. string.pack("<f", config.values[2].val) -- value 1: 4-byte float

-- Write the wire payload
local f = io.open("wire_payload.bin", "wb")
f:write(string.pack("B", MANIFEST_VERSION))
f:write(string.pack("B", #config.topology))
f:write(config.topology)
f:write(blob)
f:close()
print("Gateway: Sent config to MCU.")

File: gateway.zig
// gateway.zig - EdgeTG config author in Zig
// Compile: zig build-exe gateway.zig -O ReleaseFast
// Run: ./gateway

const std = @import("std");

// The human-readable config as a struct
const Config = struct {
    topology: []const u8 = "_/_\\",
    sensor_id: []const u8 = "sensor_01",
    temp: f32 = 25.3,
};

pub fn main() !void {
    const config = Config{};
    
    // Open wire payload file
    var file = try std.fs.cwd().createFile("wire_payload.bin", .{});
    defer file.close();
    var writer = file.writer();
    
    // 1. Write manifest version (1 byte)
    try writer.writeInt(u8, 2);
    
    // 2. Write topology length (1 byte) + topology string
    try writer.writeInt(u8, @intCast(config.topology.len));
    try writer.writeAll(config.topology);
    
    // 3. Write values blob (4-byte string len + string + 4-byte float)
    try writer.writeIntLittle(u32, @intCast(config.sensor_id.len));
    try writer.writeAll(config.sensor_id);
    try writer.writeIntLittle(f32, config.temp);
    
    std.debug.print("Gateway (Zig): Sent config to MCU.\n", .{});
}
